### ISSUE TITLE
Regression introduced in `Not` behavior

### ISSUE THREAD

**Issue by k0ral** [nikita-volkov/refined#34]
The below (simplified) code:
- raises an exception with `refined-0.4`
- doesn't raise any exception with `refined-0.3.0.0`

```haskell
{-# LANGUAGE OverloadedStrings      #-}
import           Data.Text.Prettyprint.Doc
import           Data.Text as Text
import           Data.Typeable
import           Refined

-- 'Predicate' on 'Text', true iff text is null.
data Null deriving(Typeable)

instance Predicate Null Text where
  validate p value = unless (Text.null value) $ throwRefine $ RefineOtherException (typeOf p) $ "Text is not null: " <> pretty value

foo :: Refined (Not Null) Text
foo = refineThrow "bar"
```

The exception is:
```
The negation of the predicate (Not Null) does not hold:
  The predicate (Null) does not hold: 
  Text is not null: bar
```

I suspect changeset https://github.com/nikita-volkov/refined/commit/3b9a527dfb875917244d4a8398e6c2e6b914d1e0 to be responsible for this change of behavior: 
```diff
@@ -362,8 +363,9 @@ data Not p
 instance (Predicate p x, Typeable p) => Predicate (Not p) x where
   validate p x = do
     result <- runRefineT (validate @p undefined x)
-    when (isRight result) $ do
-      throwRefine (RefineNotException (typeOf p))
+    case result of
+      Left r -> throwRefine (RefineNotException (typeOf p) r)
+      Right () -> pure ()
```
The semantics of `Right` and `Left` have been inverted. Was that intended, or is that a mistake ?

Note: I am using GHC 8.6.4.

**Comment by chessai** [comment]
That's a pretty serious regression. I apologise for screwing up like that. Unfortunately i won't be able to fix it until tomorrow, but when I do I'll make a release and mark refined-0.4 as broken.

**Comment by chessai** [comment]
see #37

**Comment by k0ral** [comment]
Thank you for fixing it quickly :) .
The behavior looks correct now, but the error message still looks strange (below examples are for an empty string):
```
The negation of the predicate (Not Null) does not hold:
```
I would rather expect:
```
The negation of the predicate (Null) does not hold:
```

**Comment by chessai** [comment]
ah, you're correct.

how would something like

```txt
The negation of the predicate (Null) does not hold:
  The subject <value_here> satisfies (Null).
```

look to you?

**Comment by k0ral** [comment]
It would look correct/accurate to me :) .

**Comment by chessai** [comment]
actually, the value cannot be produced here - we have no Pretty/Show constraint, and in general we cannot demand this, so i'm not inclined to produce a message containing the value. another reasoning behind this is that we can't know how large (by number of characters) a value is going to be, so printing it could be wasteful.

The error message could just be

```txt
The negation of the predicate (Null) does not hold.
```

Is this acceptable? (Note I'm not sure we can do much/any better in the case of `Not`)

**Comment by k0ral** [comment]
Well, error reporting is quite valuable, I would prefer if we could avoid giving it up.

Maybe the `DefaultSignatures` extension of GHC could help defining a specific implementation of `displayException` for types that are instances of `Pretty` or `Show`, but [multiple default implementation are not allowed][0], and I have the feeling we would be using a GHC extension for something it wasn't designed for.

Another idea would be to store the value within the exception type, to give users the opportunity to retrieve it and print it when catching the exception:
```haskell
data RefineException a = -- (...)

getValue :: RefineException a -> a
```
But I don't know if this would fit well with the rest of your library code -- what do you think ?

[0]: https://www.reddit.com/r/haskell/comments/8h2m24/neat_trick_to_enable_multiple_default_instance/

**Comment by chessai** [comment]
I've thought about the inclusion of the value in RefineException but
consider it to be the wrong choice. Exception types typically shouldn't
carry the value that was involved in producing the exception, since that
value can be retrieved by the user. This is evident by the type

```haskell
a -> Either e a
```

I don't want to make the choice for users, imposing on them that the value
be kept around, since the value could be potentially very large.

Most (all?) RefineExceptions defined by instances in this library never
actually show the value, only some characteristic of the value relevant to
that which we were computing. (If they do show the values, this should be
patched).

Basically, I think the user should be responsible for handling the value in
the event of a refinement failure, since the API/internals do not suit this
very well.

**Comment by k0ral** [comment]
I agree that keeping the entire value inside the exception type looks wrong. Looking at how parsing libraries dealt with that design decision, they usually store some "position" information about the original value ; but here, there doesn't seem to be any such thing as a "position" known by `refined`, this concept probably has to be handled by the user, outside of `refined`.

Let me try another way.
The only information that is stored about the value is the `validate` implementation, which essentially returns a `Either RefineException ()`. That information is asymmetric:
- `RefineException` can include a user-defined description of the value ; this is what I did with:
  ```haskell
  RefineOtherException (typeOf p) $ "Text is not null: " <> pretty value
  ```
- `()` stores no information whatsoever

In other words, the `validate` signature is biased: it makes it possible to describe error cases, but not success cases. The `Not` operator essentially swaps those cases, which is the fundamental issue here.
Hence the following suggestion:
```haskell
class (Typeable p) => Predicate p x where
  validate :: (Monad m) => p -> x -> RefineT m (Doc Void)  -- Return a string even when this succeeds

instance Predicate Null Text where
  validate p value =
    if Text.null value
    then return $ "Text is null: " <> pretty value
    else throwRefine $ RefineOtherException (typeOf p) $ "Text is not null: " <> pretty value
```
What do you think ?
